### PR TITLE
Create A Box

### PR DESCRIPTION
**Pull request by mcgnly**
This is a pretty big implementation, so I'll include a lot of screenshots. If you want to run it locally, it goes together with the branch (create-box-mutation)[https://github.com/boxwise/boxwise-flask/pull/25] which you will also have to run locally at the same time. 
Step 1: sign in and start on the home screen. I usually test with the user@user.co account, but you can use whatever. Go to "scan" with the icon indicated.
image
Step 2: select that you want to make a box
image
Step 3: scan the QR code
image
Step 4: click the button to go to the create-a-box page
image
Step 5: fill out the form, some default values have been provided for ease of testing, but you can put other things if you want. On the list for future improvements is a searchable dropdown kind of situation for the product/size, but for now, we have accepted the in-between solution of just typing in a valid id for them. See also my note about commenting in and out the QR barcode line for testing locally, it will let you use any random QR code rather than having to create a new QR code on the locally-running backend (which we don't have FE functionality for yet)
image
Step 6: Click the button to do the mutation, and if successful, you'll get a message about the box_id which you have created. You can now go check that that is now in the DB using that ID to select
image

### CODE REVIEW

File: src/views/CreateBox/CreateBox.tsx [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    variables: { email },
+  });
+
+  const [createBoxMutation, { loading: mutationLoading, error: mutationError }] = useMutation(
+    CREATE_BOX,
+  );
+
+  const location: LocationState = useLocation();
+  const qrUrl: string = location?.state?.qr;
+  // FOR TESTING
```
Comment (mcgnly) [inline src/views/CreateBox/CreateBox.tsx]: Here I have included a valid QR barcode so you don't have to generate one. If you're testing this out locally, you can comment out the top qrBarcode and uncomment the bottom one, and then it won't complain about whatever random qr code not being in the database.

File: src/views/Home.tsx [hunk 2]
```diff
@@ -8,26 +7,13 @@ function Home() {
...
-          </Link>
-          <br />
-          <Link className="text-blue-900 hover:bg-blue-200" to="/org/all">
-            {`->`} List All Bases (everyone logged in can do this)
-          </Link>
-          <br />
-          <Link className="text-blue-900 hover:bg-blue-200" to="/pdf">
-            {`->`} Generate QR Codes
-          </Link>
+          <h4>find boxes</h4>
```
Comment (mcgnly) [inline src/views/Home.tsx]: these will be styled later, I just wanted to take out my demo links because they were creating confusion

**Comment by mcgnly** [comment]
I pulled in master and it had a lot of style changes, so my screenshots may not correspond exactly anymore, sorry

**Comment by mc18g13** [comment]
@mcgnly 

Im getting this...

```
./src/index.js

Cannot find module: 'semantic-ui-less/semantic.less'. Make sure this package is installed.

You can install this package by running: yarn add semantic-ui-less/semantic.less.
```

doesn't seem to be the same on master. I am wondering if you could take a look.

**Comment by jamescrowley** [comment]
@mcgnly I have a few general suggestions as you mentioned it's a big-ish PR, I hope you don't mind :) I have to admit I first looked at this and then didn't review as realized I needed more time to do it justice. I've personally found that while making the PR simpler/smaller takes a little more time, it tends to pay back with better feedback and the faster reviews I tend to get. The screen shots and walkthrough are definitely a great idea!

It may be worth considering 'rebasing' and tidying up the commit history, so each commit is a smaller number of meaningful changes so you can 'tell the story'. The rebase gives you a chance to 'rewrite' the story so it's clearer, as invariably the actual work is more a mix of discovery and figuring out how to approach! I realise you might just have run out of time, but: https://getaround.tech/git-rebase-edit-history/ and http://dev.to/maxwell_dev/the-git-rebase-introduction-i-wish-id-had might be of interest if you haven't tried it before.

If I've understood the PR correctly, you have:

* Linting rule changes, and some files changed as a result. 
* Some changes to auth (the setAuthObject and AuthObjectType changes)
* And then the actual changes for creating a box.

I realise these all 'build' on each other, so there's a bit of a trade-off here. You could put them all in separate PRs (if they can truly stand-alone, which (1) and (2) might be able to in this case? and then flag (3) as being dependent on (1) and (2)). They would then be super easy to review/merge. Or if that doesn't feel good, perhaps just break them into their own commits, as part of a rebase ^^.

Please forgive the distraction from the PR itself!

**Review by HaGuesto — CHANGES_REQUESTED** [review]
## Functional review of Create a Box
@mcgnly @aerinsol I'm going through the different steps and test all the buttons and actions you can do. I will mention all the points I would like to be changed before the release, but this might not change the fact that we will merge this PR to remove the dependence on this branch.

#### Step 1: 
- works and my camera is turned on. very cool. 
- In my first test I did not login with user@user.co, but one of the [accounts given in the README](https://github.com/boxwise/boxwise-react#development-database-seed)
- The "Scan" button on the bottom left, does not always fire an action. Reproduce: 1. Click the "Scan" button, 2. Click the "Go Home" button, 3. Click the "Scan" button again. 

#### Step 2: 
- From a user standpoint, I do not get why the camera is turned on and it does not recognize the QR-code and at the same time there are two buttons to "make a box" or "move or edit a box".
- At a later stage, I would love to get rid of step 2 completly except there is a major reason why you made this step @mcgnly ? One should first scan the code. If a box is already connected to this QR code then it should go to the "Move or Edit the box" window. If no box is connected to the QR-code it should automatically go to the "Create a Box" window.
- The "logout" button is not visible anymore. The page is not scrollable. 

#### Step 3: 
- I can scan qr-codes and it does recognize them. Very cool :)
- At this point I would love to have the check if the hash exists in the database --> if the QR code was not generated by Boxtribute an error message should show maybe in a pop-up. Also not now but for the future.

#### Step 4:
Same as Step 2 I would like to get rid of this step completly except there is a major reason why you made this step @mcgnly ? This is just another click at the moment.

#### Step 5
- The form does not have line breaks at the moment which took me a bit to realize.
- It only offers me the location "Shop Lesvos" when I am logged in at the BoxAid organisation and it gives me "LOST Lesvos" and "SCRAP Lesvos when I am logged in as the BoxCare organisation. 
Therefore, I guess @mcgnly you already created a query to the API to return locations, but you maybe mixed up the concept of warehouse locations and bases. 
An organisation (Drop in the Ocean) can have multiple bases (Lesvos, Athens, ..). Each base can have multiple locations (Warehouse 1, warehouse 2, ...) where they store their goods from the base. 
At the moment I guess you fill the list with locations of the id of the organisation, but what you have to do is to fill the list with locations of the *base_id* of the base the user works in at the moment.
This point should be resolved now.

### Step 6
does not work for me. I get an error message. I'm logged in as `dev_coordinator@boxcare.org` from [here](https://github.com/boxwise/boxwise-react#development-database-seed). I also commented line 30 in `CreateBox.tsx` and used the QR-code you sent me. 
It does work if I use the fixed test hash in `CreateBox.tsx`

image

#### Step 7
- I can create multiple boxes with the same QR hash. I guess you did not implement this during the dev process to not always create new hashes or re-seed the database, but we need to implement this. This need to be resolved in the future.
- The `Go Home` button is almost  not visible and reachable on my screen. This need to be resolved in the future.
image

@mcgnly as a summary, I would like you to resolve the comments about step 5 and answer the questions in this comment directed at you before we merge.
Let me know if you have any questions!

**Comment by HaGuesto** [comment]
## Functional review of Auth 
The Logout button does not work at the moment. 
Reproduce:
1. Click "Log Out" 
2. Click "Sign In" 
3. It should send you to the Auth0 login page, but it does not. It directly goes to the home screen.

@mcgnly we probably need to send a request to Auth0 to delete the cache or something. What do you think?

This should be a new ticket and will not affect if we merge this PR or not.

@aerinsol we also need a consistent position of the "Log Out" button on all the screens. At the moment it is all over the place.

**Comment by HaGuesto** [comment]
## Database Review of Create a Box
I took a look at the database to check which boxes got created and see if the qr-code @mcgnly existed in my db ... it did not. 
I then inserted the hash in the `qr`-table by hand and the error I got before between **Step 6** and **Step 7** resolved @mcgnly 

So, that's already nice that it throws an error if the qr-code does not exist in the db. 

One last thing I saw when looking at the box you created @mcgnly. The column `created_by` does not get filled. It would be great if this could get filled with the user id you get from the Auth0 token.
Would this be possible now or is this a major effort?

**Comment by aerinsol** [comment]
I'll make a card to resolve the points above.

**Comment by mcgnly** [comment]
so what I understand is that before a merge you (@HaGuesto ) would like:
- the multiple buttons removed (I'll just take out the edit/move option at this point, it's just a placeholder anyway. I'd suggest that the step to do the query to determine if the box is already created be another PR- it's totally doable but just requires expanding the scope of this even more to do it here)
- the bases changed to locations (you're right, I was unaware of the distinction, I thought base was the smallest unit)
- the "created by" field filled (it was empty in the version of the db I have, so I kept it consistently empty). I should be able to do that without a big effort.

Is that a fair summary?

**Comment by mcgnly** [comment]
@HaGuesto it looks like from the user table, I can get the usergroup which gets me the bases which gets me the locations. I don't think I know where the user is currently working, I just know where their usergroup gives them access to- so the list of locations will be the total list of locations to which they have access. Does that fullfil the requirements?

**Comment by HaGuesto** [comment]
@mcgnly quite fair summary.

> the multiple buttons removed (I'll just take out the edit/move option at this point, it's just a placeholder anyway. I'd suggest that the step to do the query to determine if the box is already created be another PR- it's totally doable but just requires expanding the scope of this even more to do it here)

I take it if you want to remove the buttons, but you do not have to do that.

> @HaGuesto it looks like from the user table, I can get the usergroup which gets me the bases which gets me the locations. I don't think I know where the user is currently working, I just know where their usergroup gives them access to- so the list of locations will be the total list of locations to which they have access. Does that fullfil the requirements?
Not quite. The `created_by` db column is connected through a FOREIGN KEY to the `cms_users` table. Only the `id` of  the `cms_users` should be inserted here. 

This should be the same id as the `user_id` you get from Auth0.

**Review by HaGuesto — APPROVED** [review]
Thank you @mcgnly for the final changes!

**Pull request merged**